### PR TITLE
Feature Admin update role 187584923

### PR DESCRIPTION
**Pull request by ndahimana154**
Finished to allow the admin to change the role of any User,
The roles are ['Admin','Seller','Buyer'] only.

The remainig task on this feature is to protect the route. and checking if the acting user is Admin truly.

### CODE REVIEW

File: src/routes/userRouter.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// Import controller
+import userControllers from "../modules/user/controller/userControllers";
+
+// Import Validations
+import {validateUpdateUserRole} from "../middlewares/validation"
+
+const userRouter = express.Router();
+
+// Update the users role
+userRouter.put("/update-role/:id",validateUpdateUserRole, userControllers.updateUserRole);
```
Comment (key-joshua) [inline src/routes/userRouter.ts:13]: Please follow the code convention we are using in our project. Firstly, we are not using Joi Validations in this way.

Comment (key-joshua) [inline src/routes/userRouter.ts:13]: Rename this endpoint from `"/update-role/:id"` to `"/admin-update-role/:id"`

File: src/modules/user/controller/userControllers.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    const updatedUser = await authRepositories.findUserByAttributes("id", req.params.id)
+    return res.status(httpStatus.OK).json({
+      status: httpStatus.OK,
+      message: "User role updated successfully",
+      new: updatedUser
+    });
+  } catch (error) {
+    return res.status(httpStatus.INTERNAL_SERVER_ERROR).json({
+      status: httpStatus.INTERNAL_SERVER_ERROR,
+      message: "An error occurred while updating the user role."
```
Comment (key-joshua) [inline src/modules/user/controller/userControllers.ts:19]: You have to return error which is coming from the server not your own customized error.

File: src/routes/userRouter.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import { Router } from "express";
+import userControllers from "../modules/user/controller/userControllers";
+import {isUserExist, validation,validateUpdateUserRole,updateUserRoleSchema} from "../middlewares/validation";
+import { statusSchema } from "../modules/user/validation/userValidations";
+
+
+const router: Router = Router()
+
+router.put("/admin-update-user-status/:id", validation(statusSchema), isUserExist, userControllers.updateUserStatus);
+router.put("/admin-update-role/:id",validation(updateUserRoleSchema),validateUpdateUserRole, userControllers.updateUserRole);
```
Comment (key-joshua) [inline src/routes/userRouter.ts:10]: Rename this schema from `updateUserRoleSchema` to `roleSchema`

Comment (key-joshua) [inline src/routes/userRouter.ts:10]: `updateUserRoleSchema` should be into the module validation ( `userValidations` ) not in global middleware.

Comment (key-joshua) [inline src/routes/userRouter.ts:10]: What the use this middleware `validateUpdateUserRole` ?

File: src/modules/user/controller/userControllers.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
+// user Controllers
+import { Request, Response } from "express";
+
+import authRepositories from "../../auth/repository/authRepositories";
+import httpStatus from "http-status";
+
+const updateUserRole = async (req: Request, res: Response) => {
+  try {
+    await authRepositories.UpdateUserByAttributes("role", req.body.role, "id", req.params.id)
+    const updatedUser = await authRepositories.findUserByAttributes("id", req.params.id)
```
Comment (key-joshua) [inline src/modules/user/controller/userControllers.ts:10]: In UpdateUserByAttributes we already returned findUserByAttribute. Why are you calling it back here.

File: src/modules/user/controller/userControllers.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+// user Controllers
+import { Request, Response } from "express";
+
+import authRepositories from "../../auth/repository/authRepositories";
+import httpStatus from "http-status";
+
+const updateUserRole = async (req: Request, res: Response) => {
+  try {
+    await authRepositories.UpdateUserByAttributes("role", req.body.role, "id", req.params.id)
```
Comment (key-joshua) [inline src/modules/user/controller/userControllers.ts:9]: This function `UpdateUserByAttributes`  should start with small letter ( `updateUserByAttributes ` ).

File: src/modules/user/controller/userControllers.ts (hunk 4 above)
Comment (key-joshua) [inline src/modules/user/controller/userControllers.ts:10]: In `UpdateUserByAttributes` we already returned `findUserByAttribute`. Why are you calling it back here ?

File: src/modules/user/controller/userControllers.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    const updatedUser = await authRepositories.findUserByAttributes("id", req.params.id)
+    return res.status(httpStatus.OK).json({
+      status: httpStatus.OK,
+      message: "User role updated successfully",
+      new: updatedUser
```
Comment (key-joshua) [inline src/modules/user/controller/userControllers.ts:14]: Please follow the code convention we are using to return the response.

File: src/modules/user/controller/userControllers.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    });
+  } catch (error) {
+    return res.status(httpStatus.INTERNAL_SERVER_ERROR).json({
+      status: httpStatus.INTERNAL_SERVER_ERROR,
+      message: error
```
Comment (key-joshua) [inline src/modules/user/controller/userControllers.ts:19]: Please follow the code convention we are using to return the response in catch.

File: src/routes/userRouter.ts (hunk 3 above)
Comment (key-joshua) [inline src/routes/userRouter.ts:10]: Firstly, you have to check if a user with that `id` exist before you update.

Comment (ndahimana154) [reply src/routes/userRouter.ts]: At first i was using it checking if the user exists but i removed it, as there is the one made for that.

File: README.md [hunk 8]
```diff
@@ -23,26 +23,30 @@ This is the backend for E-Commerce-Ninjas, written in Node.js with TypeScript.
...
 
-## Completed Features
+## COMPLETED FEATURES
 
 - Welcome Endpoint
 - Register Endpoint
 - Verification Email Endpoint
 - Resend verification Endpoint
+- Login Endpoint
+- Admin change status Endpoint
```
Comment (AimePazzo) [inline README.md:34]: You didn't add your completed feature to the list.

**Review by AimePazzo — COMMENTED** [review]
Remove Public folder
image

File: package.json [hunk 9]
```diff
@@ -19,6 +19,7 @@
     "start": "ts-node-dev src/index.ts",
     "dev": "ts-node-dev src/index.ts",
     "test": "cross-env NODE_ENV=test npm run deleteAllTables && cross-env NODE_ENV=test npm run createAllTables && cross-env NODE_ENV=test npm run createAllSeeders && nyc cross-env NODE_ENV=test mocha --require ts-node/register 'src/**/*.spec.ts' --timeout 600000 --exit",
+    "test-dev": "cross-env NODE_ENV=development npm run deleteAllTables && cross-env NODE_ENV=development npm run createAllTables && cross-env NODE_ENV=development npm run createAllSeeders && nyc mocha --require ts-node/register './src/**/*.spec.ts' --timeout 300000 --exit",
```
[the remark this answers is not in the document]
Comment (ndahimana154) [reply package.json:22]: Removed

**Pull request merged**